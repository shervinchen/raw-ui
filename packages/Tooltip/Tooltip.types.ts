import { HTMLAttributes, MutableRefObject, ReactNode } from 'react';

export type TooltipPlacement =
  | 'top'
  | 'topLeft'
  | 'topRight'
  | 'left'
  | 'leftTop'
  | 'leftBottom'
  | 'bottom'
  | 'bottomLeft'
  | 'bottomRight'
  | 'right'
  | 'rightTop'
  | 'rightBottom';

export interface TooltipPosition {
  top: number;
  left: number;
  transform: string;
}

export interface TooltipArrowPosition {
  top: string;
  left: string;
  right: string;
  bottom: string;
  transform: string;
}

interface BaseTooltipProps {
  content: ReactNode;
  placement?: TooltipPlacement;
  hideArrow?: boolean;
  disabled?: boolean;
  className?: string;
  getPopupContainer?: () => HTMLElement | null;
}

type NativeTooltipProps = Omit<
  HTMLAttributes<HTMLDivElement>,
  keyof BaseTooltipProps
>;

export type TooltipProps = BaseTooltipProps & NativeTooltipProps;

export interface TooltipArrowProps {
  placement: TooltipPlacement;
  targetRef: MutableRefObject<HTMLElement | null>;
}

export interface TooltipArrowOffset {
  x: string;
  y: string;
}
