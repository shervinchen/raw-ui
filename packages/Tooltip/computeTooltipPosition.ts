import { MutableRefObject } from 'react';
import { computePopupPosition } from '../Popup/computePopupPosition';
import {
  TooltipArrowOffset,
  TooltipArrowPosition,
  TooltipPlacement,
  TooltipPosition,
} from './Tooltip.types';

export const getTargetRect = (
  targetRef?: MutableRefObject<HTMLElement | null>
) => {
  const targetRect = targetRef?.current?.getBoundingClientRect() ?? null;
  const width = targetRect
    ? targetRect.width || targetRect.right - targetRect.left
    : 0;
  const height = targetRect
    ? targetRect.height || targetRect.bottom - targetRect.top
    : 0;

  return {
    width,
    height,
  };
};

export const computeTooltipPosition = (
  placement: TooltipPlacement,
  targetRef?: MutableRefObject<HTMLElement | null>,
  getPopupContainer?: () => HTMLElement | null
) => {
  const { top, bottom, left, right } = computePopupPosition(
    targetRef,
    getPopupContainer
  );
  const { width, height } = getTargetRect(targetRef);
  const offset = 6;

  const placements: {
    [key in TooltipPlacement]: TooltipPosition;
  } = {
    top: {
      top: top - offset,
      left: left + width / 2,
      transform: 'translate(-50%, -100%)',
    },
    topLeft: {
      top: top - offset,
      left,
      transform: 'translate(0, -100%)',
    },
    topRight: {
      top: top - offset,
      left: left + width,
      transform: 'translate(-100%, -100%)',
    },
    bottom: {
      top: bottom + offset,
      left: left + width / 2,
      transform: 'translate(-50%, 0)',
    },
    bottomLeft: {
      top: bottom + offset,
      left,
      transform: 'translate(0, 0)',
    },
    bottomRight: {
      top: bottom + offset,
      left: left + width,
      transform: 'translate(-100%, 0)',
    },
    left: {
      top: top + height / 2,
      left: left - offset,
      transform: 'translate(-100%, -50%)',
    },
    leftTop: {
      top,
      left: left - offset,
      transform: 'translate(-100%, 0)',
    },
    leftBottom: {
      top: top + height,
      left: left - offset,
      transform: 'translate(-100%, -100%)',
    },
    right: {
      top: top + height / 2,
      left: right + offset,
      transform: 'translate(0, -50%)',
    },
    rightTop: {
      top,
      left: right + offset,
      transform: 'translate(0, 0)',
    },
    rightBottom: {
      top: top + height,
      left: right + offset,
      transform: 'translate(0, -100%)',
    },
  };

  return placements[placement] || placements['top'];
};

export const computeTooltipArrowPosition = (
  arrowOffset: TooltipArrowOffset,
  placement: TooltipPlacement
) => {
  const placements: {
    [key in TooltipPlacement]: TooltipArrowPosition;
  } = {
    top: {
      top: 'auto',
      right: 'auto',
      left: '50%',
      bottom: '3px',
      transform: 'translate(-50%, 100%) rotate(45deg)',
    },
    topLeft: {
      top: 'auto',
      right: 'auto',
      left: `${arrowOffset.x}`,
      bottom: '3px',
      transform: 'translate(0, 100%) rotate(45deg)',
    },
    topRight: {
      top: 'auto',
      right: `${arrowOffset.x}`,
      left: 'auto',
      bottom: '3px',
      transform: 'translate(0, 100%) rotate(45deg)',
    },
    bottom: {
      top: '3px',
      right: 'auto',
      left: '50%',
      bottom: 'auto',
      transform: 'translate(-50%, -100%) rotate(45deg)',
    },
    bottomLeft: {
      top: '3px',
      right: 'auto',
      left: `${arrowOffset.x}`,
      bottom: 'auto',
      transform: 'translate(0, -100%) rotate(45deg)',
    },
    bottomRight: {
      top: '3px',
      right: `${arrowOffset.x}`,
      left: 'auto',
      bottom: 'auto',
      transform: 'translate(0, -100%) rotate(45deg)',
    },
    left: {
      top: '50%',
      right: '3px',
      left: 'auto',
      bottom: 'auto',
      transform: 'translate(100%, -50%) rotate(45deg)',
    },
    leftTop: {
      top: `${arrowOffset.y}`,
      right: '3px',
      left: 'auto',
      bottom: 'auto',
      transform: 'translate(100%, -50%) rotate(45deg)',
    },
    leftBottom: {
      top: 'auto',
      right: '3px',
      left: 'auto',
      bottom: `${arrowOffset.y}`,
      transform: 'translate(100%, 50%) rotate(45deg)',
    },
    right: {
      top: '50%',
      right: 'auto',
      left: '3px',
      bottom: 'auto',
      transform: 'translate(-100%, -50%) rotate(45deg)',
    },
    rightTop: {
      top: `${arrowOffset.y}`,
      right: 'auto',
      left: '3px',
      bottom: 'auto',
      transform: 'translate(-100%, -50%) rotate(45deg)',
    },
    rightBottom: {
      top: 'auto',
      right: 'auto',
      left: '3px',
      bottom: `${arrowOffset.y}`,
      transform: 'translate(-100%, 50%) rotate(45deg)',
    },
  };

  return placements[placement] || placements['top'];
};
