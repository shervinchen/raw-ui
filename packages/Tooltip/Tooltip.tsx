import React, { FC, PropsWithChildren, useId, useRef, useState } from 'react';
import classNames from 'classnames';
import { TooltipProps } from './Tooltip.types';
import { useTransition } from '../utils/hooks';
import Popup from '../Popup';
import { computePopupPosition } from '../Popup/computePopup';
import TooltipArrow from './TooltipArrow';
import { useTheme } from '../Theme';

const Tooltip: FC<PropsWithChildren<TooltipProps>> = ({
  content,
  placement = 'top',
  hideArrow = false,
  disabled = false,
  className = '',
  getPopupContainer,
  children,
  ...restProps
}) => {
  const tooltipId = useId();
  const ref = useRef<HTMLDivElement>(null);
  const theme = useTheme();
  const [visible, setVisible] = useState(false);
  const { stage, shouldMount } = useTransition(visible, 0, 50);
  const classes = classNames('raw-tooltip', className);

  const handleMouseEnterOrLeave = (nextValue: boolean) => {
    if (!disabled) {
      setVisible(nextValue);
    }
  };

  return (
    <>
      <div
        data-testid="tooltipTarget"
        aria-describedby={tooltipId}
        ref={ref}
        onMouseEnter={() => handleMouseEnterOrLeave(true)}
        onMouseLeave={() => handleMouseEnterOrLeave(false)}
        className="raw-tooltip-target"
        {...restProps}
      >
        {children}
      </div>
      <Popup
        name="tooltip"
        visible={shouldMount}
        targetRef={ref}
        getPopupPosition={() =>
          computePopupPosition(placement, ref, getPopupContainer)
        }
        getPopupContainer={getPopupContainer}
      >
        <div
          role="tooltip"
          id={tooltipId}
          className={classes}
          style={{
            opacity: stage === 'enter' ? 1 : 0,
          }}
        >
          {!hideArrow && <TooltipArrow targetRef={ref} placement={placement} />}
          {content}
        </div>
      </Popup>
      <style jsx>{`
        .raw-tooltip {
          background-color: ${theme.palette.foreground};
          color: ${theme.palette.background};
          border-radius: 6px;
          font-size: 14px;
          padding: 8px 12px;
          transition: opacity 0.05s ease;
        }
        .raw-tooltip-target {
          display: inline-flex;
        }
      `}</style>
    </>
  );
};

export default Tooltip;
