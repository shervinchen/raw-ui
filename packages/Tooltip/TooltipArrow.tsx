import { FC, useMemo } from 'react';
import { TooltipArrowOffset, TooltipArrowProps } from './Tooltip.types';
import {
  computeTooltipArrowPosition,
  getTargetRect,
} from './computeTooltipPosition';
import { useTheme } from '../Theme';

const TooltipArrow: FC<TooltipArrowProps> = ({ targetRef, placement }) => {
  const theme = useTheme();
  const arrowOffset = useMemo<TooltipArrowOffset>(() => {
    const { width, height } = getTargetRect(targetRef);
    return {
      x: `${width / 2}px`,
      y: `${height / 2}px`,
    };
  }, [targetRef]);
  const { left, top, right, bottom, transform } = computeTooltipArrowPosition(
    arrowOffset,
    placement
  );

  return (
    <span>
      <style jsx>{`
        span {
          width: 6px;
          height: 6px;
          border: 0;
          background-color: ${theme.palette.foreground};
          position: absolute;
          left: ${left};
          top: ${top};
          right: ${right};
          bottom: ${bottom};
          transform: ${transform};
        }
      `}</style>
    </span>
  );
};

export default TooltipArrow;
