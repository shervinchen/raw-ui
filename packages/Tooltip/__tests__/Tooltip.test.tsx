import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import Tooltip from '..';
import userEvent from '@testing-library/user-event';

describe('Tooltip', () => {
  test('should match the snapshot', () => {
    const { asFragment } = render(
      <Tooltip content="I am a tooltip">Hover me</Tooltip>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('should support custom class name', () => {
    const { container } = render(
      <Tooltip content="I am a tooltip" className="custom-tooltip">
        Hover me
      </Tooltip>
    );
    expect(container.firstChild).toHaveClass('custom-tooltip');
  });

  test('should show tooltip when mouse over target', async () => {
    const user = userEvent.setup();
    const { getByTestId } = render(
      <Tooltip content="I am a tooltip">Hover me</Tooltip>
    );
    await user.hover(getByTestId('tooltipTarget'));
    expect(screen.getByText('I am a tooltip')).toBeInTheDocument();
  });

  test('should hide tooltip when mouse out target', async () => {
    const user = userEvent.setup();
    const { getByTestId } = render(
      <Tooltip content="I am a tooltip">Hover me</Tooltip>
    );
    const target = getByTestId('tooltipTarget');
    await user.hover(target);
    const tooltipContent = screen.getByTestId('tooltipContent');
    expect(tooltipContent).toBeInTheDocument();
    await user.unhover(target);
    await waitFor(() => {
      expect(tooltipContent).not.toBeInTheDocument();
    });
  });

  test('should support disabled tooltip', async () => {
    const user = userEvent.setup();
    const { getByTestId } = render(
      <Tooltip content="I am a tooltip" disabled>
        Hover me
      </Tooltip>
    );
    await user.hover(getByTestId('tooltipTarget'));
    await waitFor(() => {
      expect(screen.queryByTestId('tooltipContent')).not.toBeInTheDocument();
    });
  });
});
