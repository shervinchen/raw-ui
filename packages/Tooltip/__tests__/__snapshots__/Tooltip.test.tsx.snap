// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="jsx-1329010130 raw-tooltip"
    data-testid="tooltipTarget"
  >
    Hover me
  </div>
</DocumentFragment>
`;
